### PR TITLE
support hpid, huser, group

### PR DESCRIPTION
**Pull request by vrothberg**
A bunch of commits ...

### CODE REVIEW

**Comment by vrothberg** [comment]
@rhatdan if you want to start playing in advance, feel free.

**Comment by vrothberg** [comment]
Ready for review. The host data is now extract on-demand as well.

File: ps/ps.go [hunk 1]
```diff
@@ -215,6 +228,69 @@ func (p *process) cpuTime() (time.Duration, error) {
 	return cpu.Sub(time.Unix(0, 0)), nil
 }
 
+// pidsCgroupFromPID returns the path to the PID's pids cgroup.
+func pidsCgroupFromPID(pid string) (string, error) {
+	f, err := os.Open(fmt.Sprintf("/proc/%s/cgroup", pid))
```
Comment (rhatdan) [inline ps/ps.go]: Need to defer close

File: ps/ps.go [hunk 2]
```diff
@@ -215,6 +228,69 @@ func (p *process) cpuTime() (time.Duration, error) {
...
+
+// getHostProcesses returns a process slice of processes mentioned in /proc.
+// Currently, only /proc/$pid/status is parsed.
+func getHostProcesses(pid string) ([]*process, error) {
+	cgroupPath, err := pidsCgroupFromPID(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := os.Open(cgroupPath)
```
Comment (rhatdan) [inline ps/ps.go]: Need to defer close.

File: ps/ps.go [hunk 3]
```diff
@@ -445,9 +451,16 @@ func JoinNamespaceAndProcessInfo(pid, format string) ([]string, error) {
...
 	}
 
-	hostProcesses, err = getHostProcesses(pid)
-	if err != nil {
-		return nil, err
+	// extract data from host processes only on-demand / when at least one
+	// of the specified descriptors requires host data
+	for _, d := range formatDescriptors {
+		if d.onHost {
+			hostProcesses, err = getHostProcesses(pid)
```
Comment (rhatdan) [inline ps/ps.go]: Could you check if hostProcesses already set before calling this multiple times.

podman pod cid hgroup hpid huser

Will read from the system three times, I believe.

**Comment by rhatdan** [comment]
@giuseppe PTAL

File: ps/ps.go (hunk 3 above)
Comment (vrothberg) [reply ps/ps.go]: We avoid that with the `break` in line 462, so we read at most once.

File: ps/ps.go (hunk 1 above)
Comment (vrothberg) [reply ps/ps.go]: :+1:

File: ps/ps.go (hunk 2 above)
Comment (vrothberg) [reply ps/ps.go]: :+1:

File: ps/ps.go [hunk 4]
```diff
@@ -647,16 +777,60 @@ func processSECCOMP(p *process) (string, error) {
...
+// findHostProcess returns the corresponding process from `hostProcesses` for
+// provided pid or nil if non is found.
+func findHostProcess(pid string) *process {
+	for _, hp := range hostProcesses {
+		// if it's a container, the 2nd NSpid points to the
+		// corresponding PID in the container's namespace
+		if len(hp.pstatus.nSpid) < 2 {
+			continue
+		}
+		if pid == hp.pstatus.nSpid[1] {
```
Comment (giuseppe) [inline ps/ps.go]: is this check enough?  What happens if there are two containers and there are two different processes that have the same value in the second column but they are on different PID namespaces?

Comment (rhatdan) [reply ps/ps.go]: So this would be two different pids running in the same cgroup but in different pid namespaces?

Comment (vrothberg) [reply ps/ps.go]: In this case, the code is wrong. I think that using the creation of from `/proc/pid/stat` can be more useful then. To the best of my knowledge that's unique across the system as timers aren't multiplexed. WDYT?

Comment (giuseppe) [reply ps/ps.go]: @rhatdan, something like:

```console
$ unshare -rpf sleep 1000 & unshare -rpf sleep 1000 &
$ $ pgrep sleep
24122
24124
$ grep NSpid /proc/24122/status
NSpid:	24122	1
$ grep NSpid /proc/24124/status
NSpid:	24124	1
```
the two sleep processes are running in two different PID namespaces, but they still show the same `NSpid` entry from the host

Comment (giuseppe) [reply ps/ps.go]: @vrothberg I think we could just add another condition and check that the PID namespace is the same, if seen from the host or from the container PoV:

For example:
```console
$ sudo nsenter -m -t 23057 readlink -f /proc/1/ns/pid
/proc/1/ns/pid:[4026532540]

$ sudo readlink -f /proc/23057/ns/pid
/proc/23057/ns/pid:[4026532540]
```

The ID in the `[]` is the same in both cases.  That plus the code you already have should be enough IMO

**Comment by giuseppe** [comment]
@vrothberg thanks for the additional change.  LGTM!

**Comment by vrothberg** [comment]
> @vrothberg thanks for the additional change. LGTM!

Wonderful! Thanks for the great review. Test break as the readlink requires root for some processes. I'll think of a solution and will merge once tests are green.

**Comment by vrothberg** [comment]
Merging now. I'll look into code refactoring today.

**Pull request merged**